File: sdk/android-sdk/src/main/java/eu/intent/sdk/api/ITApiCallback.java
package eu.intent.sdk.api;

import android.support.annotation.NonNull;

/**
 * Implement this interface to handle the response or failure of the Intent API.
 */
public interface ITApiCallback<T> {
    /**
     * Called on request success.
     *
     * @param result the result converted into T class
     */
    void onSuccess(T result);

    /**
     * Called on request failure.
     *
     * @param httpCode  the HTTP status code, or -1 if the host was not reachable
     * @param message   a message explaining the error
     * @param errorBody the error body if provided, converted to String
     */
    void onFailure(int httpCode, @NonNull String message, @NonNull String errorBody);
}

File: sdk/android-sdk/src/main/java/eu/intent/sdk/api/ITApiCallback.kt
package eu.intent.sdk.api

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Implement this class to handle the response or failure of the Intent API.
 */
abstract class ITApiCallback<T> : Callback<T> {
    override fun onFailure(call: Call<T>?, t: Throwable?) {
        onFailure(call, ITApiError(message = t?.localizedMessage ?: ""))
    }

    override fun onResponse(call: Call<T>?, response: Response<T>?) {
        if (response == null) {
            onFailure(call, ITApiError())
        } else if (response.isSuccessful) {
            onSuccess(call, response.body())
        } else {
            val body: ITApiError = try {
                ITApiError.parseError(response)
            } catch (e: Exception) {
                ITApiError(httpCode = response.code(), httpMessage = response.message())
            }
            onFailure(call, body)
        }
    }

    /**
     * Called on request failure.
     */
    abstract fun onFailure(call: Call<T>?, body: ITApiError)

    /**
     * Called on request success.
     */
    abstract fun onSuccess(call: Call<T>?, body: T)
}